Create ForwardList.cpp

DoomzD opened the pull request.

DoomzD committed:
Create ForwardList.cpp

ifsmirnov commented on ForwardList.cpp:
```diff
@@ -0,0 +1,117 @@
+public:
+    List() : head(nullptr), sz(0) {}
+
+    void add(const int x) {
```
Лучше push_front, чтобы было понятно, куда добавляется элемент.

ifsmirnov commented on ForwardList.cpp:
```diff
@@ -0,0 +1,117 @@
+
+    void add(const int x) {
+        auto * tmp = new Node;
+        tmp->data = x;
```
Можно добавить в Node конструктор и заменить эти три строчки на одну. Или даже сейчас сделать `Node* tmp = new Node{x, head}`.

ifsmirnov commented on ForwardList.cpp:
```diff
@@ -0,0 +1,117 @@
+        ++sz;
+    }
+
+    void show() const {
```
print, output.

ifsmirnov commented on ForwardList.cpp:
```diff
@@ -0,0 +1,117 @@
+
+        Node * prev = nullptr;
+        Node * tmpNode = tmp.head;
+        while (1) {
```
while (true)

ifsmirnov commented on ForwardList.cpp:
```diff
@@ -0,0 +1,117 @@
+        Node * leftNode = left.head;
+        Node * rightNode = right.head;
+
+        while (leftNode != nullptr && rightNode != nullptr) {
```
В следующих 15 строках два блока повторяются по два раза. Можно сделать в цикле побольше ифов, но избежать лишнего копирования.

ifsmirnov commented on ForwardList.cpp:
```diff
@@ -0,0 +1,117 @@
+
+        while (leftNode != nullptr && rightNode != nullptr) {
+            if (leftNode->data < rightNode->data) {
+                tmp.add(leftNode->data);
```
add вызывает new. У тебя происходит n log n вызовов new для сортировки. Ни должно происходить ни одного.

ifsmirnov commented on ForwardList.cpp:
```diff
@@ -0,0 +1,117 @@
+            prev = this->head;
+            if (next == nullptr) break;
+            this->head = next;
+
```
Пустая строка.

ifsmirnov commented on ForwardList.cpp:
```diff
@@ -0,0 +1,117 @@
+        }
+    }
+
+    void merge(List &left, List &right) {
```
1) почему left и right принимаются по ссылке?
2) по хорошему, функция должна выглядеть как `List merge(List lhs, List rhs)`. Сейчас она записывает результат merge двух списков в текущий список. Это совсем не очевидно из сигнатуры и названия.

ifsmirnov reviewed (changes requested):
Из общих замечаний:

1) не надо делать лишних new, нужно манипулировать указателями;

2) вместо названия tmp всегда можно придумать что-то более хорошее. newNode, nextNode и так далее.

DoomzD committed:
UPD

Немного не понял, как можно уменьшить количество копирований в том блоке.
И ещё пару вопросов: 
1. Можно ли ставить такие ограничения в деструкторе?
2. Можно ли использовать названия переменных как в строках 119-120?
И да, в целом, мы придерживаемся какого-то кодстайла относительно названий переменных (не считая "говорящих" названий), или можно писать как душе угодно (всё одним стайлом конечно)?

DoomzD commented:
Немного не понял, как можно уменьшить количество копирований в том блоке.
И ещё пару вопросов: 
1. Можно ли ставить такие ограничения в деструкторе?
2. Можно ли использовать названия переменных как в строках 119-120?
И да, в целом, мы придерживаемся какого-то кодстайла относительно названий переменных (не считая "говорящих" названий), или можно писать как душе угодно (всё одним стайлом конечно)?

DoomzD committed:
UPD x2

Не ту версию отправил

ifsmirnov commented on ForwardList.cpp:
```diff
@@ -1,103 +1,143 @@
 #include <iostream>
+#include <cassert>
+
+const int BIG_INT = (1ll << 30) - 1;
```
во-первых, numeric_limits
во-вторых, нигде не используется

ifsmirnov commented on ForwardList.cpp:
```diff
@@ -1,103 +1,143 @@
+        ++sz_;
+    }
+
+    void push_front(Node* newNode) {
```
Node* node
Этот метод не должен быть публичным.

ifsmirnov commented on ForwardList.cpp:
```diff
@@ -1,103 +1,143 @@
-            Node * tmp = head->next;
-            delete head;
-            head = tmp;
+        while (head_ != nullptr) {
```
Лучше проверить isAllocated_ один раз перед циклом.

ifsmirnov commented on ForwardList.cpp:
```diff
@@ -1,103 +1,143 @@
-        Node * leftNode = left.head;
-        Node * rightNode = right.head;
+    void reverse() {
+        Node* prevNode = nullptr;
```
А вот тут Node в названии избыточно. В принципе, тут и cur нормально подходит.

ifsmirnov commented on ForwardList.cpp:
```diff
@@ -1,103 +1,143 @@
+        mergedList.reverse();
+
+        head_ = mergedList.head_;
+        Node* thisCurNode = head_;
```
Нет, это ещё хуже, чем просто cur.

ifsmirnov commented on ForwardList.cpp:
```diff
@@ -1,103 +1,143 @@
+        Node* mergedCurNode = mergedList.head_;
 
-        merge(left, right);
+        while (thisCurNode != nullptr) {
```
А зачем вообще перекидывать данные по одному? Почему нельзя просто сделать *this = merged или скопировать поля вручную?

ifsmirnov reviewed (changes requested):
Трюк с isAllocated_ мне совсем не нравится, но с ним можно смириться.

DoomzD committed:
THE FINAL UPDATE

ifsmirnov reviewed (approved):
LGTM
